DM-28583: check filters for consistency with data ID in FitsExposureFormatter

TallJimbo opened the pull request.

kfindeisen commented on python/lsst/obs/base/formatters/fitsExposure.py:
```diff
@@ -418,21 +431,98 @@ def checkUnrecognized(entry, allowed, description):
+            if (physical_filter is None and not self.dataId.hasFull()
+                    and "physical_filter" in self.dataId.graph.implied.names):
+                missing.append("physical_filter")
+        version = self._reader.readSerializationVersion()
+        if version < 2:
```
I don't understand why you have two redundant branches that depend on an implementation detail of `afw`. Either the `FilterLabel` in the `Exposure` is consistent with the data ID, or it isn't. Why do you apply completely different tests on completely different variables?

TallJimbo replied:
I did not consider the serialization version an implementation detail; I considered it the best way to find out whether the filter value(s) in the file are likely to be wrong or incomplete.  I'm happy to switch to a different approach if you can identify one.

And essentially "are the filter values in the file trustworthy" is what the branching here is about.  If the file values are likely to be wrong, we don't want to warn when they are, but we can fix them, but we do want to warn if we can't fix them.  If the file values are likely to be right, we warn if the data ID says they're wrong, and don't warn if we can't check them.

TallJimbo replied:
FWIW, with the new logic (following the discussion), there is still a notion of trustworthiness derived from the serialization version here, but a much weaker one; it's now only used to eliminate warnings for pre-`FilterLabel` files.

kfindeisen replied:
Can you explain the reasoning behind that behavior? At the very least, users should probably be warned that they're using old files -- especially since the quality of the translation will become _much_ worse once we remove `Filter`.

TallJimbo replied:
As far as I can tell, all of DP0.1 will be using old files, and iwe may have HSC files that old in use for a while yet, too.  Anyone going through this formatter who does not get a warning will not care whether they get old files; the quality of the filter translation will be irrelevant because it'll be overwritten by what's in the data ID.  (Unless I misunderstood what you meant by "translation".)

kfindeisen commented on python/lsst/obs/base/formatters/fitsExposure.py:
```diff
@@ -212,6 +215,16 @@ def readComponent(self, component, parameters=None):
                 if component == "dimensions" and thisComponent is not None:
                     thisComponent = thisComponent.getDimensions()
+
+                if component == "FilterLabel":
```
The component is `filterLabel` (lower case). Should you do a similar check for the `filter` component, or is this already covered because it's derived from `filterLabel`?

TallJimbo replied:
> The component is `filterLabel` (lower case).

:+1: And hmm, now you've got me wondering whether my tests are doing what I think they're doing.

> Should you do a similar check for the filter component, or is this already covered because it's derived from filterLabel?

I was not planning to check `Filter` at all, because I didn't think I could predict what the correct content would be from a `Filter` given the data ID (especially if "correct" might mean "what it was before filter standardization").

TallJimbo replied:
Ah, I see what happened with `FilterLabel` vs `filterLabel` - this is a relic from an earlier version, and the (now) operative check does use `filterLabel` elsewhere.  I'll remove this `if` block entirely.

parejkoj reviewed (commented):
There are no tests of the `_compareFilters` logic here. It looks like that method doesn't actually need any files, so could it be tested entirely independent of the files, no? There are a lot of `if` branches and boolean logic, which you could be sure to explicitly exercise with individual units. 

At least mention somewhere that it is tested by ci_hsc_gen3?

TallJimbo commented on python/lsst/obs/base/formatters/fitsExposure.py:
```diff
@@ -22,24 +22,27 @@
 
-class FitsExposureFormatter(Formatter):
-    """Interface for reading and writing Exposures to and from FITS files.
+class FitsImageFormatterBase(Formatter):
```
`Formatter` is itself an `ABC`, so this already is, too.

TallJimbo commented on python/lsst/obs/base/formatters/fitsExposure.py:
```diff
@@ -418,21 +422,94 @@ def checkUnrecognized(entry, allowed, description):
+
+        Parameters
+        ----------
+        file_filter_label : `lsst.afw.image.FilterLabel`, optional
```
Good point; it can be `None`, but that needs to be explicit.  Will change to "or None".

TallJimbo commented on python/lsst/obs/base/formatters/fitsExposure.py:
```diff
@@ -418,21 +422,94 @@ def checkUnrecognized(entry, allowed, description):
+            # Warn if that's more likely than not to be bad, because the file
+            # predates filter standardization.
+            if version < 2:
+                warnings.warn(f"Data ID {self.dataId} is missing (implied) value(s) for {missing}; "
```
That's a good question, and I simply don't know the answer to which should be preferred.  I feel like `warnings` is what the broader Python world is likely to use, and I know how to write tests against it, so that's what I've used here, but I hadn't noticed the use of `lsst.log` in `FitsRawFormatter`.  Seems like something @timj might have thoughts on.

timj replied:
I'm fine with this being logger rather than warnings. I obviously have a preference for `logging` rather than `lsst.log` in butler code. `pipetask` and `cmdLineTask` have to deal with both anyhow.  You can use `assertLogs` in tests if you are using python `logging`.

TallJimbo replied:
If `warnings` isn't a problem, I'd actually prefer to go with that, here, if only because

> You can use `assertLogs` in tests if you are using python `logging`.

probably evaluates to `False` here in obs_base, and I haven't see a good argument for using `lsst.log` just to warn.

parejkoj replied:
Where would those `warnings.warn` messages end up in a production environment? Are they obvious in the log output?

Also you can use `assertLogs` with `lsst.log` too, you just have to wrap it in a `with lsst.log.UsePythonLogging():`.

timj replied:
You can wrap but I'd use python logging here anyhow if I was going to use logging. Python warnings are a bit odd in general since people seem to ignore them (see all the tests that warn).

TallJimbo replied:
I've raised this on community; the dev-guide should probably say something to preempt future versions of this conversation.

I don't have strong opinions, but the community thread doesn't resolve decisively soon, I'm going to merge it as is.

TallJimbo commented on python/lsst/obs/base/formatters/fitsExposure.py:
```diff
@@ -418,21 +422,94 @@ def checkUnrecognized(entry, allowed, description):
+            warnings.warn(f"Reading {self.fileDescriptor.location} with data ID {self.dataId}: "
+                          f"filter label mismatch (file is {file_filter_label}, data ID is "
+                          f"{data_id_filter_label}).  This is probably a bug in the code that produced it.")
+        return data_id_filter_label
```
My understanding is that `FilterLabel(band=None, physical=None)` is not a valid construct, so the appropriate `FilterLabel` to attach to an `Exposure` that isn't associated with a filter (say, a bias) in a structural sense is `None`/`nullptr`.  Is that correct?

TallJimbo commented on python/lsst/obs/base/formatters/fitsExposure.py:
```diff
@@ -418,21 +422,94 @@ def checkUnrecognized(entry, allowed, description):
+
+    _readerClass = ExposureFitsReader
+
+    def _compareFilters(self, file_filter_label):
```
:+1: to `_fixFilterLabel`.  I didn't go with that originally because the originally logic _didn't_ fix it most of the time, but now it makes more sense.

TallJimbo commented on tests/test_fitsRawFormatter.py:
```diff
@@ -136,9 +136,15 @@ def setUp(self):
+        universe = lsst.daf.butler.DimensionUniverse()
+        dataId = lsst.daf.butler.DataCoordinate.standardize(instrument="Cam1", exposure=2, detector=10,
+                                                            physical_filter="Cam1-R",
+                                                            band="r", universe=universe)
```
Correct.  I think if new code needs them, they can just switch `dataId` to `self.dataId`, but it's premature to do that now.

Also, I did make `detector=10` consistent with line 90, because I don't like leaving inconsistencies lurking in test data, and I didn't notice until now that there was also some metadata that used `u` as a filter-something, so I'll switch the filters used here, too.  I didn't see anything that would specify or otherwise constrain what the other made-up values should be.

TallJimbo commented on python/lsst/obs/base/formatters/fitsExposure.py:
```diff
@@ -418,21 +422,94 @@ def checkUnrecognized(entry, allowed, description):
+        -------
+        filter_label : `lsst.afw.image.FilterLabel` or `None`
+            The preferred filter label; may be the given one or one built from
+            the data ID.
```
Given that we override the one from the file in all other circumstances where they differ, and (I think) `None` is the only way to express the state `band=None, physical=None` (a `FilterLabel` design choice I endorse!), I think we do need to return `None`.  Note that it's not just that _this_ data ID does not have those entries; it's that data IDs for this dataset type cannot ever have a band or physical_filter.  When it's just the former, we should already be returning the file's `FilterLabel` (in the `if missing` block earlier).  I'll update the docs to clarify that.

TallJimbo commented:
I think I've now addressed all review comments (including those on other PRs) except these ones:

> There are no tests of the _compareFilters logic here. It looks like that method doesn't actually need any files, so could it be tested entirely independent of the files, no? There are a lot of if branches and boolean logic, which you could be sure to explicitly exercise with individual units.

I'm afraid it does need the file, even though it's pretty subtle: it needs `self._reader`, and that can only be initialized with a file, in another method (and the fact that it is so subtle is lot of why I created DM-28698).  I think there might be a way to restructure it to make it possible to test without the file, but I'm not sure it will work, and I'd really like to get this in the weekly.  Would you comfortable signing off on the ticket with the ci_hsc_gen3 coverage that exists, with that extra testing a possible bonus that might land before I merge?

> At least mention somewhere that it is tested by ci_hsc_gen3?

Certainly.

TallJimbo commented:
Earlier than expected update on adding tests to obs_base: the restructuring I was thinking of is done, and overall it's an improvement, but overall it's not enough, I think: the `Formatter` needs the file passed to it on construction.  And while it's possible the base class doesn't do much with that file, and I could get through it with mocks, I think the fragility/encapsulation-breaking that mocking would add isn't worth the "coverage proximity" adding tests here would bring.

parejkoj commented:
I'm still a bit concerned about the total number of clauses across the different `if`s here, but if it's difficult to implement a test without having files, then i guess we'll have to live with it for now. Would DM-28698 make implementing local unittests easier?

TallJimbo commented:
> Would DM-28698 make implementing local unittests easier?

It's hard to say for certain without figuring out how we'd solve the problems that ticket raises, but the fact that it's hard to write local now is definitely closely related to those problems, so I hope so.

timj commented:
obs_base is impossible to write tests for because you don't have any data because you'd need a real obs package. We end up running tests in the other obs packages. We haven't got an obs_test for gen3 that uses tiny images with simplified metadata and a special metadata translator. It's not clear we gain anything over using obs_lsst or obs_decam or obs_subaru.

parejkoj commented:
Responding to @timj : that's the point of mocks and *unit* testing, so you don't need files, you just test the logic independent of files.

timj commented:
Re testing, I'd rather we had very simple YAML and FITS files with metadata that we ingested so as to test the whole flow rather than mocking out the butler.  Testing a Formatter is easier if a butler is involved since you are guaranteed that you are using the formatter properly (its only purpose is to be used by Butler). It's even easier to let obs_xxx downstream packages test everything and that's the reason this has not been tackled.

TallJimbo commented:
> I'd rather we had very simple YAML and FITS files with metadata that we ingested so as to test the whole flow rather than mocking out the butler.

:+1:.  I have spent way too much time fixing tests unrelated to my changes that broke because the mock _data_ they used didn't actually conform to our data model (something that is pretty hard for mock data to do), and we suddenly (and reasonably) started validating the data against that data model.  It's also true that I didn't know enough about mocks a year or two ago to use them in some tests where I should have, but I still think that it's important to distinguish between mocking APIs (which mocking libraries are good at) and mocking complex data (much harder).

Anyhow, this reminds me of something @timothydmorton recently brought up: we should be more careful about sure butler repositories are as useful as possible even if the user doesn't have the code used to create them (e.g. like the concrete obs_ package code used to register instruments and ingest raws).  There are some things we can never support, like `butler.get` without the `Formatter` used to write the dataset, and for the most part we have gotten lucky about most other things just working, but it's something we should pay more attention to, I think, to make sure putting real-world test data upstream of obs packages continues to work.
